File: haskell/src/Concurrency/ChannelThroughput.hs
module Concurrency.ChannelThroughput
  ( channelRoundtrip
  ) where

import Control.Concurrent (forkIO)
import Control.Concurrent.STM (atomically)
import Control.Concurrent.STM.TBQueue
    ( newTBQueueIO, readTBQueue, writeTBQueue )
import Control.Monad (forM_)
import Data.IORef (newIORef, readIORef, modifyIORef')
import Data.Word (Word64)

-- | Send n messages through a bounded TBQueue (STM).
-- Matches Rust's crossbeam bounded channel.
-- Returns the sum of received values.
channelRoundtrip :: Word64 -> IO Word64
channelRoundtrip n = do
    queue <- newTBQueueIO 1024
    sumRef <- newIORef (0 :: Word64)

    -- Sender thread
    _ <- forkIO $
        forM_ [0..n-1] $ \i ->
            atomically $ writeTBQueue queue i

    -- Receiver (on main thread)
    forM_ [1..n] $ \_ -> do
        val <- atomically $ readTBQueue queue
        modifyIORef' sumRef (+ val)

    readIORef sumRef

File: rust/src/concurrency/channel_throughput.rs
use crossbeam_channel::{bounded, Receiver, Sender};
use std::thread;

/// Send n messages through a bounded crossbeam channel.
/// Returns the sum of received values (forces complete consumption).
pub fn channel_roundtrip(n: u64) -> u64 {
    let (tx, rx): (Sender<u64>, Receiver<u64>) = bounded(1024);

    let sender = thread::spawn(move || {
        for i in 0..n {
            tx.send(i).expect("send failed");
        }
    });

    let receiver = thread::spawn(move || {
        let mut sum = 0u64;
        for val in rx {
            sum = sum.wrapping_add(val);
        }
        sum
    });

    sender.join().expect("sender panicked");
    receiver.join().expect("receiver panicked")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn should_receive_all_messages() {
        let sum = channel_roundtrip(100);
        let expected: u64 = (0..100).sum();
        assert_eq!(sum, expected);
    }

    #[test]
    fn should_handle_zero_messages() {
        assert_eq!(channel_roundtrip(0), 0);
    }
}
